jetty-server/src/main/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandler.java

```diff
@@ -0,0 +1,376 @@
+
+            // If we are not committed, must decide if we should aggregate or not.
+            if (_aggregating == null)
+            {
+                if (last)
+                    _aggregating = Boolean.FALSE;
+                else
+                {
+                    Response response = _channel.getResponse();
+                    int sc = response.getStatus();
+                    if (sc > 0 && (sc < 200 || sc == 204 || sc == 205 || sc >= 300))
+                        _aggregating = Boolean.FALSE;  // No body
+                    else
+                    {
+                        String ct = response.getContentType();
+                        if (ct == null)
+                            _aggregating = Boolean.TRUE;
+                        else
+                        {
+                            ct = MimeTypes.getContentTypeWithoutCharset(ct);
+                            _aggregating = isMimeTypeBufferable(StringUtil.asciiToLowerCase(ct));
+                        }
+                    }
+                }
+            }
+
+            // If we are not aggregating, then handle normally.
+            if (!_aggregating)
+            {
+                getNextInterceptor().write(content, last, callback);
+                return;
+            }
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} aggregating", this);
+
+            try
+            {
+                aggregate(content);
+            }
+            catch (Throwable t)
+            {
+                resetBuffer();
+                callback.failed(t);
+                return;
+            }
+
+            if (last)
+                commit(callback);
+            else
+                callback.succeeded();
+        }
+
+        protected void aggregate(ByteBuffer content) throws IOException
+        {
+            if (_bufferedOutputStream == null)
+            {
+                // Create a new OutputStream to a file.
+                _file = Files.createTempFile(tempDir, "BufferedResponse", "").toFile();
+                _bufferedOutputStream = new BufferedOutputStream(Files.newOutputStream(_file.toPath(), StandardOpenOption.WRITE));
+            }
+
+            BufferUtil.writeTo(content, _bufferedOutputStream);
+        }
+
+        @Override
+        public Interceptor getNextInterceptor()
+        {
+            return _next;
+        }
+
+        protected void commit(Callback callback)
+        {
+            try
+            {
+                _bufferedOutputStream.flush();
+                _bufferedOutputStream.close();
+                _bufferedOutputStream = null;
+            }
+            catch (IOException e)
+            {
+                resetBuffer();
+                callback.failed(e);
+                return;
+            }
+
+            // Create an iterating callback to do the writing
+            IteratingCallback icb = new IteratingCallback()
```

Comment:
I'm not sure we need an ICB if we are using a file mapped buffer, as that will write the entire content in a single write.  So process should only be called once with _last==false???

We'd only need an ICB if a file mapped buffer is not used and we have to read,write,read,write... the content a buffer at a time.

Reply:
@greg the max size of ByteBuffer is 2GB so if the file is bigger than that we can't have the whole file as a single MappedByteBuffer, so I decided to use the Iterating callback. I could use `Integer.MAX_VALUE` instead of `OutputBufferSize` for the size of the buffers, but I think we would still need the ICB.